standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java

```diff
@@ -9322,9 +9334,25 @@ public void processContext(ServerContext serverContext, TTransport tTransport, T
     if (startLock != null) {
       signalOtherThreadsToStart(tServer, startLock, startCondition, startedServing);
     }
+
+    // If dynamic service discovery through ZooKeeper is enabled, add this server to the ZooKeeper.
+    if (MetastoreConf.getVar(conf, ConfVars.THRIFT_SERVICE_DISCOVERY_MODE)
+            .equalsIgnoreCase("zookeeper")) {
+      zooKeeperHelper = MetastoreConf.getZKConfig(conf);
+      String serverInstanceURI = getServerInstanceURI(port);
+      zooKeeperHelper.addServerInstanceToZooKeeper(serverInstanceURI, serverInstanceURI, null,
+              new ZKDeRegisterWatcher(zooKeeperHelper));
+      HMSHandler.LOG.info("Metastore server instance with URL " + serverInstanceURI + " added to " +
+              "the zookeeper");
+    }
+
     tServer.serve();
   }
 
+  private static String getServerInstanceURI(int port) throws Exception {
+    return ServerUtils.getHostAddress(null).getHostAddress() + ":" + port;
```

Comment:
using the raw ip address might not be a good way of doing it, in case there is dual network card etc. might be better to use a fqdn and even with that there can be different names associated might be better to have a config to get the address to provide in the zookeeper discover url for this hms. From the code looks like HMS binds to * for now